Prevent h1 smuggling (#177)
ietf-wg-masque/draft-ietf-masque-connect-ip: draft-ietf-masque-connect-ip.md

Additionally, IPv6 stateless autoconfiguration or router advertisement messages
are not used in such interfaces, and neither is neighbor discovery.

> When using HTTP/2 or later, a client MAY optimistically start sending proxied IP packets before receiving

Reviewer: Wondering if "later" is the correct thing here. Perhaps:
```suggestion
When using HTTP/2 or HTTP/3, or future HTTP versions, a client MAY optimistically start sending proxied IP packets before receiving
```

Reviewer 2: This sounds like we are introducing a design contract that applies to all the future versions of HTTP. Are we fine with that?

Author: Agreed, switched to h2 and h3

a single end-to-end flow. Because of this, such use of DSCPs in
privacy-sensitive contexts is NOT RECOMMENDED.

Opportunistic sending of IP packets (see {{link-operation}}) is only allowed in
versions of HTTP that are greater or equal to 2, because, in HTTP/1.1, a server
that does not implement this specification could reject the HTTP Upgrade and

> attempt to parse the IP packets as HTTP message content or as a subsequent HTTP

Reviewer 2: > as HTTP message content

Is this possible? IIRC connect-ip sends neither Content-Length nor Transfer-encoding, and that makes the HTTP/1.1 server assume that the size of the request body to be zero, assuming that it ignored the upgrade. Therefore, I think all the opportunistic payload will be considered as "subsequent HTTP requests"?

Author: Good point, removed message content and kept subsequent requests

> Opportunistic sending of IP packets (see {{link-operation}}) is only allowed in
> versions of HTTP that are greater or equal to 2, because, in HTTP/1.1, a server
> that does not implement this specification could reject the HTTP Upgrade and
> attempt to parse the IP packets as HTTP message content or as a subsequent HTTP
> request, allowing request smuggling attacks; see
> {{?OPTIMISTIC=I-D.schwartz-httpbis-optimistic-upgrade}}.

Reviewer: ```suggestion
Opportunistic sending of IP packets (see {{link-operation}}) is not allowed in
version prior to HTTP/2 because, in HTTP/1.1, a server that does not implement
this specification could reject the HTTP Upgrade and attempt to parse the IP
packets as HTTP message content or as a subsequent HTTP request, allowing
request smuggling attacks; see
{{?OPTIMISTIC=I-D.schwartz-httpbis-optimistic-upgrade}}.
```
